here-naksha-handler-activitylog/src/test/java/com/here/naksha/handler/activitylog/ActivityLogHandlerTest.java

```diff
@@ -1,205 +1,451 @@
+        .hasChildrenThat(
+            first -> first
+                .hasType(POpType.EQ)
+                .hasPRef(PRef.uuid()),
+            second -> second
+                .hasType(POpType.EQ)
+                .hasPRef(PRef.id())
+        );
   }
 
   @Test
-  void test_fromActivityLogPartial() throws IOException {
-    final XyzFeature feature =
-        JsonSerializable.deserialize(IoHelp.openResource("activity_log_partial.json"), XyzFeature.class);
-    assertNotNull(feature);
-    assertNotNull(feature.getProperties());
-    assertNotNull(feature.getProperties().getXyzNamespace());
-    assertNotNull(feature.getProperties().getXyzActivityLog());
-    assertNotNull(feature.getProperties().getXyzActivityLog().getOriginal());
-    final String xyzNamespacePuuid =
-        feature.getProperties().getXyzActivityLog().getOriginal().getPuuid();
-    final String xyzNamespaceMuuid =
-        feature.getProperties().getXyzActivityLog().getOriginal().getMuuid();
-    final String xyzNamespaceSpace =
-        feature.getProperties().getXyzActivityLog().getOriginal().getSpace();
-    final long xyzNamespaceCreatedAt =
-        feature.getProperties().getXyzActivityLog().getOriginal().getCreatedAt();
-    final long xyzNamespaceUpdatedAt =
-        feature.getProperties().getXyzActivityLog().getOriginal().getUpdatedAt();
-    ActivityLogHandler.fromActivityLogFormat(feature);
-    assertSame(xyzNamespacePuuid, feature.getProperties().getXyzNamespace().getPuuid());
-    assertSame(xyzNamespaceMuuid, feature.getProperties().getXyzNamespace().getMuuid());
-    assertSame(xyzNamespaceSpace, feature.getProperties().getXyzNamespace().getSpace());
-    assertEquals(
-        xyzNamespaceCreatedAt, feature.getProperties().getXyzNamespace().getCreatedAt());
-    assertEquals(
-        xyzNamespaceUpdatedAt, feature.getProperties().getXyzNamespace().getUpdatedAt());
-    assertNull(feature.getProperties().getXyzActivityLog());
+  void shouldComposeActivityFeatures() throws Exception {
+    // Given
+    String featureId = "featureId";
+    XyzFeature oldFeature = xyzFeature(
+        featureId,
+        "initial_uuid",
+        null,
+        EXyzAction.CREATE,
+        Map.of(
+            "op", "old feature",
+            "magicNumber", 123
+        )
+    );
+    XyzFeature newFeature = xyzFeature(
+        featureId,
+        "new_uuid",
+        "initial_uuid",
+        EXyzAction.UPDATE,
+        Map.of(
+            "op", "new feature",
+            "magicBoolean", true
+        )
+    );
+
+    // And
+    ReadFeatures request = new ReadFeatures();
+    spaceStorageSessionReturningHistoryFeatures(request, oldFeature, newFeature);
+
+    // When
+    Result result = handler.processEvent(eventWith(request));
+
+    // Then:
+    assertThatResult(result)
+        .hasActivityFeatures(
+            firstFeature -> firstFeature
+                .hasId(newFeature.getProperties().getXyzNamespace().getUuid())
```

Comment:
IMO we would be testing wrong layers then - it's not job of the handler to serialize the entire result, this handler might be used by a client who will be using `XyzFeature` objects, not serialized ones. 

If this was REST API test then I would totally agree. That being said - I understand this might get difficult to read. I'll try to work on that.